[CI] Python CI is unstable because of downloading bazel error

### Search before asking

- [x] I had searched in the [issues](https://github.com/apache/fory/issues) and found no similar issues.


### Version

Latest

### Component(s)

Other

### Description
Python CI randomly fails sometimes.

### Minimal reproduce step

submit a pr

### What did you expect to see?

ci pass

### What did you see instead?
log: [Python CI (3.13.3, ubuntu-24.04-arm)](https://github.com/apache/fory/actions/runs/18349720040/job/52266048481?pr=2724#logs)
```
Running command Getting requirements to build editable
  2025/10/08 15:25:24 Downloading https://releases.bazel.build/6.3.2/release/bazel-6.3.2-linux-arm64...
  2025/10/08 15:25:29 could not download Bazel: failed to download bazel: failed to download bazel: could not copy from https://releases.bazel.build/6.3.2/release/bazel-6.3.2-linux-arm64 to /home/runner/.cache/bazelisk/downloads/_tmp/download1245162363: stream error: stream ID 1; INTERNAL_ERROR; received from peer
  DEBUG = False, BAZEL_BUILD_EXT = True, PATH = /tmp/pip-build-env-8vv6bk_o/overlay/bin:/tmp/pip-build-env-8vv6bk_o/normal/bin:/opt/hostedtoolcache/Python/3.13.3/arm64/bin:/opt/hostedtoolcache/Python/3.13.3/arm64:/snap/bin:/home/runner/.local/bin:/opt/pipx_bin:/home/runner/.cargo/bin:/home/runner/.config/composer/vendor/bin:/opt/google-cloud-sdk/bin:/home/runner/.dotnet/tools:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin
  setup_dir: /home/runner/work/fory/fory/python
  project_dir: /home/runner/work/fory/fory
  fory_cpp_src_dir: /home/runner/work/fory/fory/src
  Traceback (most recent call last):
    File "/opt/hostedtoolcache/Python/3.13.3/arm64/lib/python3.13/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 389, in <module>
      main()
      ~~~~^^
    File "/opt/hostedtoolcache/Python/3.13.3/arm64/lib/python3.13/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 373, in main
      json_out["return_val"] = hook(**hook_input["kwargs"])
                               ~~~~^^^^^^^^^^^^^^^^^^^^^^^^
    File "/opt/hostedtoolcache/Python/3.13.3/arm64/lib/python3.13/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 157, in get_requires_for_build_editable
      return hook(config_settings)
    File "/tmp/pip-build-env-8vv6bk_o/overlay/lib/python3.13/site-packages/setuptools/build_meta.py", line 473, in get_requires_for_build_editable
      return self.get_requires_for_build_wheel(config_settings)
             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^
    File "/tmp/pip-build-env-8vv6bk_o/overlay/lib/python3.13/site-packages/setuptools/build_meta.py", line 331, in get_requires_for_build_wheel
      return self._get_build_requires(config_settings, requirements=[])
             ~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "/tmp/pip-build-env-8vv6bk_o/overlay/lib/python3.13/site-packages/setuptools/build_meta.py", line 301, in _get_build_requires
      self.run_setup()
      ~~~~~~~~~~~~~~^^
    File "/tmp/pip-build-env-8vv6bk_o/overlay/lib/python3.13/site-packages/setuptools/build_meta.py", line 317, in run_setup
      exec(code, locals())
      ~~~~^^^^^^^^^^^^^^^^
    File "<string>", line 64, in <module>
    File "/tmp/pip-build-env-8vv6bk_o/overlay/lib/python3.13/site-packages/setuptools/__init__.py", line 115, in setup
      return distutils.core.setup(**attrs)
             ~~~~~~~~~~~~~~~~~~~~^^^^^^^^^
    File "/tmp/pip-build-env-8vv6bk_o/overlay/lib/python3.13/site-packages/setuptools/_distutils/core.py", line 148, in setup
      _setup_distribution = dist = klass(attrs)
                                   ~~~~~^^^^^^^
    File "<string>", line 57, in __init__
    File "/opt/hostedtoolcache/Python/3.13.3/arm64/lib/python3.13/subprocess.py", line 419, in check_call
      raise CalledProcessError(retcode, cmd)
  subprocess.CalledProcessError: Command '['bazel', 'build', '-s', '--copt=-fsigned-char', '//:cp_fory_so']' returned non-zero exit status 1.
  error: subprocess-exited-with-error
  
  × Getting requirements to build editable did not run successfully.
  │ exit code: 1
  ╰─> See above for output.
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  full command: /opt/hostedtoolcache/Python/3.13.3/arm64/bin/python /opt/hostedtoolcache/Python/3.13.3/arm64/lib/python3.13/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py get_requires_for_build_editable /tmp/tmpzt54lry5
  cwd: /home/runner/work/fory/fory/python
  Getting requirements to build editable: finished with status 'error'
error: subprocess-exited-with-error

× Getting requirements to build editable did not run successfully.
│ exit code: 1
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.
Error: Process completed with exit code 1.
```

### Anything Else?

_No response_

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

--- a/ci/run_ci.sh
+++ b/ci/run_ci.sh
@@ -96,7 +96,23 @@ install_bazel() {
   BINARY_URL="https://github.com/bazelbuild/bazel/releases/download/${BAZEL_VERSION}/bazel-${BAZEL_VERSION}-${OS}-${ARCH}"
 
   echo "Downloading bazel from: $BINARY_URL"
-  curl -L -sSf -o "$BAZEL_DIR/bazel" "$BINARY_URL" || { echo "Failed to download bazel"; exit 1; }
+  # Retry download with exponential backoff to avoid transient network errors in CI
+  MAX_ATTEMPTS=5
+  SLEEP_SECONDS=2
+  attempt=1
+  while [ $attempt -le $MAX_ATTEMPTS ]; do
+    if curl -L -sSf -o "$BAZEL_DIR/bazel" "$BINARY_URL"; then
+      break
+    fi
+    echo "Attempt $attempt to download bazel failed."
+    if [ $attempt -eq $MAX_ATTEMPTS ]; then
+      echo "Failed to download bazel after $MAX_ATTEMPTS attempts"
+      exit 1
+    fi
+    sleep $SLEEP_SECONDS
+    SLEEP_SECONDS=$((SLEEP_SECONDS * 2))
+    attempt=$((attempt + 1))
+  done
   chmod +x "$BAZEL_DIR/bazel"
 
   # Add to current shell's PATH

--- a/ci/tasks/common.py
+++ b/ci/tasks/common.py
@@ -15,7 +15,6 @@
 # specific language governing permissions and limitations
 # under the License.
 
-import shutil
 import subprocess
 import platform
 import urllib.request as ulib
@@ -88,6 +87,34 @@ def get_bazel_download_url():
     return f"{download_url_base}/bazel-{bazel_version}-{get_os_name_lower()}-{get_os_machine()}"
 
 
+def urlretrieve_with_retries(url, filename, max_attempts=5, initial_delay=2):
+    """Download a URL to filename with retries and exponential backoff.
+
+    Raises the last exception if all attempts fail.
+    """
+    attempt = 1
+    delay = initial_delay
+    last_exc = None
+    while attempt <= max_attempts:
+        try:
+            logging.info(f"Downloading (attempt {attempt}) {url} -> {filename}")
+            ulib.urlretrieve(url, filename)
+            return
+        except Exception as e:
+            logging.error(f"Download attempt {attempt} failed: {e}")
+            last_exc = e
+            if attempt == max_attempts:
+                break
+            logging.info(f"Retrying in {delay} seconds...")
+            import time
+
+            time.sleep(delay)
+            delay *= 2
+            attempt += 1
+    logging.error(f"All {max_attempts} download attempts failed for URL: {url}")
+    raise last_exc
+
+
 def cd_project_subdir(subdir):
     """Change to a subdirectory of the project."""
     os.chdir(os.path.join(PROJECT_ROOT_DIR, subdir))
@@ -124,7 +151,7 @@ def install_bazel():
         # For Windows, download the installer and add it to PATH
         local_name = "bazel.exe"
         try:
-            ulib.urlretrieve(bazel_download_url, local_name)
+            urlretrieve_with_retries(bazel_download_url, local_name)
         except Exception as e:
             logging.error(f"Failed to download bazel: {e}")
             logging.error(f"URL: {bazel_download_url}")
@@ -142,7 +169,7 @@ def install_bazel():
         bazel_path = os.path.join(home_bin, "bazel")
 
         try:
-            ulib.urlretrieve(bazel_download_url, bazel_path)
+            urlretrieve_with_retries(bazel_download_url, bazel_path)
         except Exception as e:
             logging.error(f"Failed to download bazel: {e}")
             logging.error(f"URL: {bazel_download_url}")

